shared/src/PayloadCollection.cs

```diff
@@ -0,0 +1,154 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.Azure.Devices.Shared
+{
+    /// <summary>
+    /// The base class for all payloads that accept a <see cref="PayloadConvention"/>
+    /// </summary>
+    public abstract class PayloadCollection : IEnumerable<object>
+    {
+        /// <summary>
+        /// The underlying collection for the payload.
+        /// </summary>
+        public IDictionary<string, object> Collection { get; private set; } = new Dictionary<string, object>();
+
+        /// <summary>
+        /// The convention to use with this payload.
+        /// </summary>
+        public PayloadConvention Convention { get; internal set; }
```

Comment:
one drawback of placing these common resources in the shared assembly is that either all APIs need to be public (even though they shouldn't be, design-wise), OR we need to add the client library as a "friend" to the shared library (InternalsVisibleTo).

The other, not-so-elegant approach is to add these common resources to the "common" folder and add a compile-include condition of the client csproj files.

What do you think?
- make all methods public
- add the device client assembly as friend to shared assembly - allow InternalsVisibleTo
- add the common files as compile-include to the project

Reply:
Why does it need to be in shared?

Reply:
The idea is that since these are IoT hub defined conventions, they could be relevant for both device and service side applications, hence, could be relevant for both device clients and service clients.
While we do have very pnp specific APIs on the service client, the "regular" service client and registry manager also can also be used to perform operations on pnp-enabled devices. If we are to ever make those APIs follow these conventions, that would be the motivation for placing these in shared.

Reply:
So we're creating a challenge for ourselves by trying to future proof for a problem we don't know we'll have?

Reply:
That is absolutely correct, we are future-proofing ourselves for an issue that we don't know for sure we'll have, but since these operations have both device and service side components, it does seem reasonable that we might want to extend it at some point.

Reply:
Until we know these specific types, operating the exact same way as they do on the device, are useful to the service, I'd suggest we do not put them in the shared library. Just put them in device. Before we ship GA, if you find a compelling reason to share them, we can address it then.

YAGNI